Keycloak - configuring for production on Codesphere

sancho1952007 opened the pull request:
# Keycloak - configuring for production on Codesphere

## Links
- Blog Post: https://medium.com/@sanchogodinho/keycloak-configuring-for-production-on-codesphere-b1bd9dd79901

## About Keycloak
Keycloak is an open-source identity and access management solution that provides robust authentication and authorization services for web and mobile applications. Developed by Red Hat, Keycloak simplifies the implementation of security features by offering a centralized and customizable authentication server. It supports various authentication protocols, including OpenID Connect, OAuth 2.0, and SAML, allowing developers to integrate single sign-on (SSO) capabilities seamlessly. Keycloak's versatile features include user federation, role-based access control, and social identity provider integration, enabling organizations to secure their applications with ease. Its user-friendly administration console facilitates the management of users, groups, and client applications, making it a popular choice for enterprises seeking a comprehensive identity management solution.

## What I achieved
- Tutorial on creating a database using Codesphere.
- Setting up the workspace for the installation of Keycloak.
- Scripts to stall Keycloak on Codesphere in production mode.

## Checklist
- [x] My project has been thoroughly tested and is fully functional.
- [x] The repository for my project includes a `ci.yml` file (Refer to [CodeSphere CI Pipelines](https://docs.codesphere.com/getting-started/ci-pipelines)).
- [x] The blog post about my project has been published.
- [x] All workspaces for this project are deployed on Codesphere using the CI pipeline.
- [x] My app is persistently installed in a workspace (all components are installed and configured somewhere below the `/home/user/app` path).
- [x] My pull request is well-written and clear.

Datata1 commented:
Hey Sancho,

Thank you for your contribution! The template and the blog post are looking good so far! 👍

However, there are three minor things that need to be changed before we can merge it:

In your blog post, you didn't mention the .sh script of the template. It would be interesting to have a section in your blog post where you explain what this shell script does in the CI pipeline. You can copy the code of the .sh file to your blog post inside three backticks (```) so it is formatted like code.

```sh
#!/bin/bash

# Check for required environment variables
required_vars=("KEYCLOAK_ADMIN" "KEYCLOAK_ADMIN_PASSWORD" "DATABASE_URL")
missing_vars=()

if [ -f ./KEYCLOAK_VERSION ]; then
    echo [LOG] Removing Previous Version Of Keycloak!
    rm -rf keycloak-$(cat ./KEYCLOAK_VERSION)
fi

for var in "${required_vars[@]}"; do
    if [ -z "${!var}" ]; then
        missing_vars+=("$var")
    fi
done

if [ ${#missing_vars[@]} -gt 0 ]; then
    echo "Please set the following environment variables before installation: ${missing_vars[*]}"
    exit 1
fi

# Proceed with installation
repo_url="https://github.com/keycloak/keycloak"
version_file="KEYCLOAK_VERSION"
latest_tag=$(curl -sSLI -o /dev/null -w '%{url_effective}' "${repo_url}/releases/latest" | awk -F/ '/tag/{print $NF}')

if [ -z "$latest_tag" ]; then
    echo "Failed to retrieve the latest release tag. Exiting."
    exit 1
fi

echo "$latest_tag" > "$version_file"
download_url="${repo_url}/releases/download/${latest_tag}/keycloak-${latest_tag}.tar.gz"
destination_file="keycloak-${latest_tag}.tar.gz"

echo "[LOG] Downloading Latest Version Of Keycloak: ${download_url}..."
curl -sSL -o "${destination_file}" "${download_url}" && echo "Download successful! File saved as ${destination_file}" || echo "Download failed. Please try again."

echo "[LOG] Installing Java via Zulu"
nix-env -iA nixpkgs.zulu

echo "[LOG] Extracting Keycloak"
tar -xvf ./keycloak-${latest_tag}.tar.gz

echo "[LOG] Deleting the tar file"
rm -rf ./keycloak-${latest_tag}.tar.gz

echo "[LOG] Installation Completed. Execute the \"RUN\" CI command."
```
- Two options are missing to set up a production-ready Keycloak instance: a) Limit the number of queued requests, and b) decide whether IPv4 or IPv6 is used.

You can find instructions on how to set up these two options here: [link to Keycloak configuration for production](https://www.keycloak.org/server/configuration-production).

sancho1952007 commented:
> Hey Sancho,
> 
> Thank you for your contribution! The template and the blog post are looking good so far! 👍
> 
> However, there are three minor things that need to be changed before we can merge it:
> 
> In your blog post, you didn't mention the .sh script of the template. It would be interesting to have a section in your blog post where you explain what this shell script does in the CI pipeline. You can copy the code of the .sh file to your blog post inside three backticks (```) so it is formatted like code.
> 
> ```sh
> #!/bin/bash
> 
> # Check for required environment variables
> required_vars=("KEYCLOAK_ADMIN" "KEYCLOAK_ADMIN_PASSWORD" "DATABASE_URL")
> missing_vars=()
> 
> if [ -f ./KEYCLOAK_VERSION ]; then
>     echo [LOG] Removing Previous Version Of Keycloak!
>     rm -rf keycloak-$(cat ./KEYCLOAK_VERSION)
> fi
> 
> for var in "${required_vars[@]}"; do
>     if [ -z "${!var}" ]; then
>         missing_vars+=("$var")
>     fi
> done
> 
> if [ ${#missing_vars[@]} -gt 0 ]; then
>     echo "Please set the following environment variables before installation: ${missing_vars[*]}"
>     exit 1
> fi
> 
> # Proceed with installation
> repo_url="https://github.com/keycloak/keycloak"
> version_file="KEYCLOAK_VERSION"
> latest_tag=$(curl -sSLI -o /dev/null -w '%{url_effective}' "${repo_url}/releases/latest" | awk -F/ '/tag/{print $NF}')
> 
> if [ -z "$latest_tag" ]; then
>     echo "Failed to retrieve the latest release tag. Exiting."
>     exit 1
> fi
> 
> echo "$latest_tag" > "$version_file"
> download_url="${repo_url}/releases/download/${latest_tag}/keycloak-${latest_tag}.tar.gz"
> destination_file="keycloak-${latest_tag}.tar.gz"
> 
> echo "[LOG] Downloading Latest Version Of Keycloak: ${download_url}..."
> curl -sSL -o "${destination_file}" "${download_url}" && echo "Download successful! File saved as ${destination_file}" || echo "Download failed. Please try again."
> 
> echo "[LOG] Installing Java via Zulu"
> nix-env -iA nixpkgs.zulu
> 
> echo "[LOG] Extracting Keycloak"
> tar -xvf ./keycloak-${latest_tag}.tar.gz
> 
> echo "[LOG] Deleting the tar file"
> rm -rf ./keycloak-${latest_tag}.tar.gz
> 
> echo "[LOG] Installation Completed. Execute the \"RUN\" CI command."
> ```
> - Two options are missing to set up a production-ready Keycloak instance: a) Limit the number of queued requests, and b) decide whether IPv4 or IPv6 is used.
> 
> You can find instructions on how to set up these two options here: [link to Keycloak configuration for production](https://www.keycloak.org/server/configuration-production).
> 
> 

Sure, will work on it!

Simon-Count commented on Keycloak/config.sh:
```diff
@@ -0,0 +1,28 @@
+rm -rf ./keycloak-$(cat ./KEYCLOAK_VERSION)/conf/keycloak.conf
+
+# Writing a new, updated configuration
+echo "db=postgres" > ./keycloak-$(cat ./KEYCLOAK_VERSION)/conf/keycloak.conf
```
Let's not write the username, password and database URL to a file as people might accidentally commit these secrets then but instead export them as environment variable with `export DB_USERNAME="$username"`. 

We can then reference these in the ci.yml run stage like so:
--db-username=$DB_USERNAME

Simon-Count replied:
Did you see my suggestions in Discord as to how the env vars could be passed in here?

Simon-Count commented on Keycloak/ci.yml:
```diff
@@ -0,0 +1,12 @@
+run:
+  steps:
+    - name: Start Keycloak
+      command: bash ./keycloak-$(cat ./KEYCLOAK_VERSION)/bin/kc.sh start --hostname $WORKSPACE_DEV_DOMAIN --http-port 3000 --proxy-headers xforwarded --http-enabled true
```
Let's add the secrets as command line arguments from env vars here - see comment below for an example.

Simon-Count reviewed (changes requested):
Thanks Sancho, one small suggestion that will improve the security and eliminate the possibility of people accidentally pushing secrets written to a file during CI. 

Can you work on these suggestions? Then we should be good to merge it! No worries about the timeline, we are still well within a reasonable timeframe.

sancho1952007 commented:
Hey, I added a .gitignore file to prevent pushing the sensitive content to git.

I thank Simon Pfeiffer for the suggestion since environmental variables weren't working that well in my case.

Simon-Count commented:
> Hey, I added a .gitignore file to prevent pushing the sensitive content to git.
> 
> I thank Simon Pfeiffer for the suggestion since environmental variables weren't working that well in my case.

@sancho1952007  Please verify whether you pushed the commit :) There is no .gitignore file in the keycloak template folder

sancho1952007 commented:
> > Hey, I added a .gitignore file to prevent pushing the sensitive content to git.
> > 
> > I thank Simon Pfeiffer for the suggestion since environmental variables weren't working that well in my case.
> 
> @sancho1952007  Please verify whether you pushed the commit :) There is no .gitignore file in the keycloak template folder

I have added it to the config.sh file. The config.sh will generate it based on the keycloak version used. You can try it by updating your existing config.sh file with the new one I committed.

I did this because my script installs the latest version of keycloak to get the latest security updates & features. If I make a static .gitignore file, it may only protect the config file for the current version that's why I added it in such a way that it will dynamically put the latest keycloak version installed to the .gitignore file.

The .gitignore file will automatically be created when you use the Prepare CI Command no extra steps needed.

Simon-Count commented:
> > > Hey, I added a .gitignore file to prevent pushing the sensitive content to git.
> > > I thank Simon Pfeiffer for the suggestion since environmental variables weren't working that well in my case.
> > 
> > 
> > @sancho1952007  Please verify whether you pushed the commit :) There is no .gitignore file in the keycloak template folder
> 
> I have added it to the config.sh file. The config.sh will generate it based on the keycloak version used. You can try it by updating your existing config.sh file with the new one I committed.
> 
> I did this because my script installs the latest version of keycloak to get the latest security updates & features. If I make a static .gitignore file, it may only protect the config file for the current version that's why I added it in such a way that it will dynamically put the latest keycloak version installed to the .gitignore file.
> 
> The .gitignore file will automatically be created when you use the Prepare CI Command no extra steps needed.

That seems unnecessarily obscure. It's always better to be literal whenever possible. And in this case I think we can just add a static .gitignore for the entire keycloak folder and use a wildcard for the version number. See here for an explanation of .gitignore wildcards: https://git-scm.com/docs/gitignore

Simon-Count reviewed (approved):
Looks good thanks!
